extensions/librdkafka/ConsumeKafka.cpp

```diff
@@ -0,0 +1,553 @@
+    if (!message || RD_KAFKA_RESP_ERR_NO_ERROR != message->err) {
+      break;
+    }
+    utils::print_kafka_message(message, logger_);
+    messages.emplace_back(std::move(message), utils::rd_kafka_message_deleter());
+    elapsed = std::chrono::high_resolution_clock::now() - start;
+  }
+  return messages;
+}
+
+utils::KafkaEncoding ConsumeKafka::key_attr_encoding_attr_to_enum() const {
+  if (utils::StringUtils::equalsIgnoreCase(key_attribute_encoding_, KEY_ATTR_ENCODING_UTF_8)) {
+    return utils::KafkaEncoding::UTF8;
+  }
+  if (utils::StringUtils::equalsIgnoreCase(key_attribute_encoding_, KEY_ATTR_ENCODING_HEX)) {
+    return utils::KafkaEncoding::HEX;
+  }
+  throw minifi::Exception(ExceptionType::PROCESSOR_EXCEPTION, "\"Key Attribute Encoding\" property not recognized.");
+}
+
+utils::KafkaEncoding ConsumeKafka::message_header_encoding_attr_to_enum() const {
+  if (utils::StringUtils::equalsIgnoreCase(message_header_encoding_, MSG_HEADER_ENCODING_UTF_8)) {
+    return utils::KafkaEncoding::UTF8;
+  }
+  if (utils::StringUtils::equalsIgnoreCase(message_header_encoding_, MSG_HEADER_ENCODING_HEX)) {
+    return utils::KafkaEncoding::HEX;
+  }
+  throw minifi::Exception(ExceptionType::PROCESSOR_EXCEPTION, "Key Attribute Encoding property not recognized.");
+}
+
+std::string ConsumeKafka::resolve_duplicate_headers(const std::vector<std::string>& matching_headers) const {
+  if (MSG_HEADER_KEEP_FIRST == duplicate_header_handling_) {
+    return matching_headers.front();
+  }
+  if (MSG_HEADER_KEEP_LATEST == duplicate_header_handling_) {
+    return matching_headers.back();
+  }
+  if (MSG_HEADER_COMMA_SEPARATED_MERGE == duplicate_header_handling_) {
+    return utils::StringUtils::join(", ", matching_headers);
+  }
+  throw minifi::Exception(ExceptionType::PROCESSOR_EXCEPTION, "\"Duplicate Header Handling\" property not recognized.");
+}
+
+std::vector<std::string> ConsumeKafka::get_matching_headers(const rd_kafka_message_t* message, const std::string& header_name) const {
+  // Headers fetched this way are freed when rd_kafka_message_destroy is called
+  // Detaching them using rd_kafka_message_detach_headers does not seem to work
+  rd_kafka_headers_t* headers_raw;
+  const rd_kafka_resp_err_t get_header_response = rd_kafka_message_headers(message, &headers_raw);
+  if (RD_KAFKA_RESP_ERR__NOENT == get_header_response) {
+    return {};
+  }
+  if (RD_KAFKA_RESP_ERR_NO_ERROR != get_header_response) {
+    logger_->log_error("Failed to fetch message headers: %d: %s", rd_kafka_last_error(), rd_kafka_err2str(rd_kafka_last_error()));
+  }
+  std::vector<std::string> matching_headers;
+  for (std::size_t header_idx = 0;; ++header_idx) {
+    const char* value;  // Not to be freed
+    std::size_t size;
+    if (RD_KAFKA_RESP_ERR_NO_ERROR != rd_kafka_header_get(headers_raw, header_idx, header_name.c_str(), (const void**)(&value), &size)) {
+      break;
+    }
+    if (size < std::numeric_limits<int>::max()) {
```

Comment:
Should this be an `assert` or a `gsl_Expects` or similar instead of an `if`?  If the length of a header is more than 2 GB, then something must have gone wrong.

Reply:
I am not sure if there is something inherently wrong with having long messages, except that we cannot reliably log them. It *might* be a use-case.

Reply:
These are headers, not messages.  Is a single header larger than 2 GB a valid use case?

Reply:
Added a 200 character limit on the logs.